File: mcp/src/app-link.ts
import net from "node:net";
import { randomBytes } from "node:crypto";
import { FrameCodec, InvalidFrame, deriveKeys } from "./crypto.js";
import { adbForward } from "./adb.js";
import type { Config } from "./config.js";
import type { Link } from "./link.js";

const PROTO_VERSION = 2;

/**
 * Connection to the on-device app over a loopback tunnel.
 *
 * There is no relay and no cloud. The app listens on `127.0.0.1:<port>` on the
 * phone; `adb forward` maps a port on this machine onto it; this link dials that.
 * So the "network" is a USB cable or an adb-over-Wi-Fi session the user
 * established, and the sealed frames never leave the pair of machines.
 *
 * Requests are strictly single-flight. The wire protocol carries a sequence
 * number but no correlation id, so responses are matched to requests by
 * ordering, which holds only while exactly one request is outstanding. The
 * device dispatches serially regardless.
 *
 * Each connection derives fresh keys from a random salt sent in the opening
 * hello, so the sequence counter starts at zero every time without any risk of
 * replaying a nonce under a reused key.
 */
export class AppLink implements Link {
  private sock: net.Socket | null = null;
  private codec: FrameCodec | null = null;
  private readonly aad: Buffer;
  private queue: Promise<unknown> = Promise.resolve();
  private pending: ((value: unknown) => void) | null = null;
  private buf = "";

  constructor(private readonly config: Config) {
    this.aad = Buffer.from(config.deviceId, "utf8");
  }

  async send(request: unknown, timeoutMs = 30_000): Promise<any> {
    const run = async () => {
      await this.ensureConnected();
      return this.exchange(request, timeoutMs);
    };

    // Chain onto the queue, keeping it alive through failures so one rejected
    // call does not poison every subsequent one.
    const result = this.queue.then(run, run);
    this.queue = result.catch(() => undefined);
    return result;
  }

  close(): void {
    this.sock?.destroy();
    this.sock = null;
    this.codec = null;
  }

  private async ensureConnected(timeoutMs = 15_000): Promise<void> {
    if (this.sock && !this.sock.destroyed && this.codec) return;

    // Bridge the device's loopback port to ours. Harmless when already mapped.
    await adbForward(this.config.port, this.config.port);

    // Fresh per-connection salt and keys. The device folds the same salt in when
    // it reads the hello, so both sides land on the same directional keys.
    const salt = randomBytes(16);
    const keys = deriveKeys(
      this.config.serverPrivateKey,
      this.config.devicePublicKey,
      salt,
    );
    const codec = new FrameCodec(keys.serverToDevice, keys.deviceToServer);

    await new Promise<void>((resolve, reject) => {
      const sock = net.connect({ host: "127.0.0.1", port: this.config.port });

      const timer = setTimeout(() => {
        sock.destroy();
        reject(new Error(`connection timed out after ${timeoutMs}ms`));
      }, timeoutMs);

      sock.once("error", (err) => {
        clearTimeout(timer);
        reject(err);
      });

      sock.once("connect", () => {
        clearTimeout(timer);
        sock.setNoDelay(true);
        sock.write(
          JSON.stringify({ v: PROTO_VERSION, salt: salt.toString("base64url") }) + "\n",
        );
        this.sock = sock;
        this.codec = codec;
        this.buf = "";
        this.attach(sock);
        resolve();
      });
    });
  }

  private attach(sock: net.Socket): void {
    sock.on("data", (chunk: Buffer) => {
      this.buf += chunk.toString("utf8");
      let idx: number;
      while ((idx = this.buf.indexOf("\n")) >= 0) {
        const line = this.buf.slice(0, idx);
        this.buf = this.buf.slice(idx + 1);
        if (!line) continue;

        let env: { deviceId: string; seq: number | string; payload: string };
        try {
          env = JSON.parse(line);
        } catch {
          continue;
        }
        if (env.deviceId !== this.config.deviceId) continue;

        let plaintext: Buffer;
        try {
          plaintext = this.codec!.open(BigInt(env.seq), env.payload, this.aad);
        } catch (e) {
          // Unopenable frames are noise or an attempt; do not answer, do not
          // surface. Only the paired device can produce a valid one.
          if (!(e instanceof InvalidFrame)) throw e;
          continue;
        }

        const resolve = this.pending;
        this.pending = null;
        resolve?.(JSON.parse(plaintext.toString()));
      }
    });

    sock.on("close", () => {
      this.sock = null;
      this.codec = null;
      this.buf = "";
      const resolve = this.pending;
      this.pending = null;
      resolve?.({ ok: false, error: "connection closed" });
    });

    // Errors surface through 'close'; swallow so they do not become unhandled.
    sock.on("error", () => undefined);
  }

  private exchange(request: unknown, timeoutMs: number): Promise<any> {
    const sock = this.sock;
    const codec = this.codec;
    if (!sock || !codec) throw new Error("not connected");

    const sealed = codec.seal(Buffer.from(JSON.stringify(request), "utf8"), this.aad);

    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        this.pending = null;
        resolve({ ok: false, error: `device did not respond within ${timeoutMs}ms` });
      }, timeoutMs);

      this.pending = (value) => {
        clearTimeout(timer);
        resolve(value);
      };

      // seq as a JSON number: the device decodes it straight into a Long.
      sock.write(
        JSON.stringify({
          deviceId: this.config.deviceId,
          seq: Number(sealed.seq),
          payload: sealed.payload,
        }) + "\n",
      );
    });
  }
}

File: desktop/src/main/kotlin/com/hereliesaz/mcpserved/desktop/net/AppLink.kt
package com.hereliesaz.mcpserved.desktop.net

import com.hereliesaz.mcpserved.desktop.adb.Adb
import com.hereliesaz.mcpserved.desktop.config.Config
import com.hereliesaz.mcpserved.desktop.crypto.Crypto
import com.hereliesaz.mcpserved.desktop.crypto.FrameCodec
import com.hereliesaz.mcpserved.desktop.crypto.InvalidFrame
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.security.SecureRandom
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * How to reach the on-device control server.
 *
 * Three shapes: an `adb forward` tunnel onto the device's loopback port (the
 * USB / paired-cable path), a direct LAN socket to an address the device
 * advertised over mDNS (the auto-discovery path), or — opt-in, off the local
 * network entirely — a relay room the device dialed out to
 * ([RemoteRelayClient] on the device side; see `relay/README.md`). Either way
 * the sealed-frame crypto is what authenticates the peer; the transport
 * underneath is just plumbing, which is why only [connect] differs between
 * them and every other line of [AppLink] is shape-agnostic.
 */
sealed class Target {
    abstract val describe: String

    /** Opens the underlying [FrameStream] for this target. */
    abstract fun connect(timeoutMs: Int): FrameStream

    /** The classic path: bridge the device's loopback port through adb. */
    data class Loopback(val port: Int) : Target() {
        override val describe get() = "adb-forward → 127.0.0.1:$port"
        override fun connect(timeoutMs: Int): FrameStream {
            Adb.forward(port, port) // harmless when already mapped
            return connectSocket("127.0.0.1", port, timeoutMs)
        }
    }

    /** A device found on the LAN over mDNS — dial it straight, no adb. */
    data class Lan(val host: String, val port: Int) : Target() {
        override val describe get() = "lan $host:$port"
        override fun connect(timeoutMs: Int): FrameStream = connectSocket(host, port, timeoutMs)
    }

    /**
     * A relay room, dialed as the "host" role. Requires no local network path
     * to the device at all — the device found the same relay independently
     * (opt-in, see `app/.../grant/RemoteAccessStore.kt`) and dialed the same
     * room as the "device" role.
     */
    data class Relay(val url: String, val room: String) : Target() {
        override val describe get() = "relay $url"
        override fun connect(timeoutMs: Int): FrameStream = RelayFrameStream(url, room)
    }

    companion object {
        fun loopback(config: Config): Target = Loopback(config.port)
        fun lan(host: String, port: Int): Target = Lan(host, port)
        fun relay(url: String, room: String): Target = Relay(url, room)
    }
}

/**
 * Connection to the on-device app over a loopback tunnel or a discovered LAN
 * address.
 *
 * There is no relay and no cloud. The app listens on the phone; this link dials
 * it — either through an `adb forward` map onto its loopback port, or directly at
 * the address it broadcast over mDNS. The sealed frames never leave the pair of
 * machines.
 *
 * Requests are strictly single-flight. The wire protocol carries a sequence
 * number but no correlation id, so responses match requests by ordering, which
 * holds only while exactly one request is outstanding. A process-wide lock
 * enforces that; the device dispatches serially regardless.
 *
 * Each connection derives fresh keys from a random salt sent in the opening
 * hello, so the sequence counter starts at zero every time without any risk of
 * replaying a nonce under a reused key.
 */
class AppLink(
    private val config: Config,
    private val target: Target,
) : Link {

    private val aad: ByteArray = config.deviceId.toByteArray(Charsets.UTF_8)
    private val lock = ReentrantLock()

    private var stream: FrameStream? = null
    private var codec: FrameCodec? = null

    override val label: String get() = "app (${target.describe})"

    override fun send(request: JsonObject, timeoutMs: Long): JsonObject = lock.withLock {
        try {
            ensureConnected()
            exchange(request, timeoutMs)
        } catch (e: Exception) {
            close()
            err(e.message ?: e.toString())
        }
    }

    override fun close() = lock.withLock {
        runCatching { stream?.close() }
        stream = null; codec = null
    }

    private fun ensureConnected(timeoutMs: Int = 15_000) {
        if (stream != null && codec != null) return

        // Fresh per-connection salt and keys. The device folds the same salt in
        // when it reads the hello, so both sides land on the same directional keys.
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val keys = Crypto.deriveKeys(config.serverPrivateKey, config.devicePublicKey, salt)
        val newCodec = FrameCodec(sealKey = keys.serverToDevice, openKey = keys.deviceToServer)

        val newStream = target.connect(timeoutMs)
        val hello = buildJsonObject {
            put("v", JsonPrimitive(PROTO_VERSION))
            put("salt", JsonPrimitive(Crypto.b64Url(salt)))
        }
        newStream.writeLine(ProtoJson.encodeToString(JsonObject.serializer(), hello))

        stream = newStream
        codec = newCodec
    }

    private fun exchange(request: JsonObject, timeoutMs: Long): JsonObject {
        val stream = stream ?: return err("not connected")
        val codec = codec ?: return err("not connected")

        val sealed = codec.seal(
            ProtoJson.encodeToString(JsonObject.serializer(), request).toByteArray(Charsets.UTF_8),
            aad,
        )
        // seq as a JSON number: the device decodes it straight into a Long.
        val envelope = buildJsonObject {
            put("deviceId", JsonPrimitive(config.deviceId))
            put("seq", JsonPrimitive(sealed.seq))
            put("payload", JsonPrimitive(sealed.payloadB64))
        }
        stream.writeLine(ProtoJson.encodeToString(JsonObject.serializer(), envelope))

        val deadline = System.currentTimeMillis() + timeoutMs
        while (true) {
            val remaining = deadline - System.currentTimeMillis()
            if (remaining <= 0) return err("device did not respond within ${timeoutMs}ms")
            val line = stream.readLine(remaining) ?: return err("connection closed")
            if (line.isEmpty()) continue

            val env = try {
                ProtoJson.parseToJsonElement(line) as? JsonObject ?: continue
            } catch (_: Exception) {
                continue
            }
            if (env.str("deviceId") != config.deviceId) continue
            val seq = env.str("seq")?.toLongOrNull() ?: continue
            val payload = env.str("payload") ?: continue

            val plaintext = try {
                codec.open(seq, payload, aad)
            } catch (_: InvalidFrame) {
                // Unopenable frames are noise or an attempt; skip them. Only the
                // paired device can produce a valid one.
                continue
            }
            return ProtoJson.parseToJsonElement(String(plaintext, Charsets.UTF_8)) as? JsonObject
                ?: err("device returned a non-object response")
        }
    }

    private companion object {
        const val PROTO_VERSION = 2
    }
}
